storage: control per-replica timestamp cache memory

The timestamp cache is currently per-replica and is the most significant user of per-replica memory. #9102 fixed an issue where we weren't clearing the timestamp cache when we lost the range lease, but even with that change the timestamp cache consumes a significant chunk of memory that grows with the number of replicas. 

One options is we could periodically trim the timestamp cache on idle replicas. At the moment, we only trim the timestamp cache when it is accessed. This would be a straightforward change yet it would interact with the feature that allows the `timestampCache` to hold very old entries as long as it hasn't hit its size threshold which makes interactive SQL shell usage nicer.

Alternately, we could move to a per-store timestamp cache. See discussion in #9101. The tricky part here is dealing with setting the low-water mark on lease acquisition. It looks possible to replace using `TimestampCache.SetLowWater` with inserting a span that covers the entire key range for which a lease is being acquired. A per-store timestamp cache is likely preferable if it could be made to work.

Cc @cockroachdb/stability

@bdarnell I'm getting test failures with my per-Store timestamp cache change which I've narrowed to the following diff:

```diff
diff --git a/pkg/storage/replica_proposal.go b/pkg/storage/replica_proposal.go
index d53c3e7..8010747 100644
--- a/pkg/storage/replica_proposal.go
+++ b/pkg/storage/replica_proposal.go
@@ -404,8 +404,17 @@ func (r *Replica) leasePostApply(
                        log.Infof(ctx, "new range lease %s following %s [physicalTime=%s]",
                                newLease, prevLease, r.store.Clock().PhysicalTime())
                }
+
+               desc := r.Desc()
                r.tsCacheMu.Lock()
-               r.tsCacheMu.cache.SetLowWater(newLease.Start)
+               for _, keyRange := range makeReplicatedKeyRanges(desc) {
+                       // TODO(peter): Figure why adding a "write" entry causes various test
+                       // failures.
+                       for _, readOnly := range []bool{true, false} {
+                               r.tsCacheMu.cache.add(
+                                       keyRange.start.Key, keyRange.end.Key, newLease.Start, nil, readOnly)
+                       }
+               }
                r.tsCacheMu.Unlock()

                // Reset the request counts used to make lease placement decisions whenever
```

Why would adding a "write" span cause:

```
~ ~/Development/go/src/github.com/cockroachdb/cockroach/pkg/storage master go test -v -run TestUnreplicateFirstRange
...
--- FAIL: TestUnreplicateFirstRange (0.18s)
	client_test.go:1057: change replicas of r1 failed: replay txn
```

--- a/pkg/storage/helpers_test.go
+++ b/pkg/storage/helpers_test.go
@@ -255,9 +255,19 @@ func (r *Replica) GetLease() (*roachpb.Lease, *roachpb.Lease) {
 
 // GetTimestampCacheLowWater returns the timestamp cache low water mark.
 func (r *Replica) GetTimestampCacheLowWater() hlc.Timestamp {
-	r.tsCacheMu.Lock()
-	defer r.tsCacheMu.Unlock()
-	return r.tsCacheMu.cache.lowWater
+	r.store.tsCacheMu.Lock()
+	defer r.store.tsCacheMu.Unlock()
+	t := r.store.tsCacheMu.cache.lowWater
+	// Bump the per-Store low-water mark using the per-range read and write info.
+	start := roachpb.Key(r.Desc().StartKey)
+	end := roachpb.Key(r.Desc().EndKey)
+	if r, _, ok := r.store.tsCacheMu.cache.GetMaxRead(start, end); !ok && t.Less(r) {
+		t = r
+	}
+	if w, _, ok := r.store.tsCacheMu.cache.GetMaxWrite(start, end); !ok && t.Less(w) {
+		t = w
+	}
+	return t
 }
 
 // GetRaftLogSize returns the raft log size.

--- a/pkg/storage/replica.go
+++ b/pkg/storage/replica.go
@@ -299,13 +299,6 @@ type Replica struct {
 		global, local *CommandQueue
 	}
 
-	tsCacheMu struct {
-		// Protects all fields in the tsCacheMu struct.
-		syncutil.Mutex
-		// Most recent timestamps for keys / key ranges.
-		cache *timestampCache
-	}
-
 	mu struct {
 		// Protects all fields in the mu struct.
 		syncutil.RWMutex
@@ -626,7 +619,6 @@ func (r *Replica) initLocked(
 	r.cmdQMu.local = NewCommandQueue(false /* !optimizeOverlap */)
 	r.cmdQMu.Unlock()
 
-	r.tsCacheMu.cache = newTimestampCache(clock)
 	r.mu.proposals = map[storagebase.CmdIDKey]*ProposalData{}
 	r.mu.checksums = map[uuid.UUID]replicaChecksum{}
 	// Clear the internal raft group in case we're being reset. Since we're
@@ -1559,9 +1551,10 @@ func (ec *endCmds) done(br *roachpb.BatchResponse, pErr *roachpb.Error, retry pr
 	// marked as affecting the cache are processed. Inconsistent reads
 	// are excluded.
 	if pErr == nil && retry == proposalNoRetry && ec.ba.ReadConsistency != roachpb.INCONSISTENT {
-		ec.repl.tsCacheMu.Lock()
-		ec.repl.tsCacheMu.cache.AddRequest(makeCacheRequest(&ec.ba, br))
-		ec.repl.tsCacheMu.Unlock()
+		creq := makeCacheRequest(&ec.ba, br)
+		ec.repl.store.tsCacheMu.Lock()
+		ec.repl.store.tsCacheMu.cache.AddRequest(creq)
+		ec.repl.store.tsCacheMu.Unlock()
 	}
 
 	ec.repl.cmdQMu.Lock()
@@ -1847,13 +1840,15 @@ func (r *Replica) beginCmds(
 // will inform the batch response timestamp or batch response txn
 // timestamp.
 func (r *Replica) applyTimestampCache(ba *roachpb.BatchRequest) (bool, *roachpb.Error) {
-	r.tsCacheMu.Lock()
-	defer r.tsCacheMu.Unlock()
+	// TODO(peter): We only need to hold a write lock during the ExpandRequests
+	// calls. Investigate whether using a RWMutex here reduces lock contention.
+	r.store.tsCacheMu.Lock()
+	defer r.store.tsCacheMu.Unlock()
 
 	if ba.Txn != nil {
-		r.tsCacheMu.cache.ExpandRequests(ba.Txn.Timestamp)
+		r.store.tsCacheMu.cache.ExpandRequests(ba.Txn.Timestamp)
 	} else {
-		r.tsCacheMu.cache.ExpandRequests(ba.Timestamp)
+		r.store.tsCacheMu.cache.ExpandRequests(ba.Timestamp)
 	}
 
 	var bumped bool
@@ -1866,7 +1861,7 @@ func (r *Replica) applyTimestampCache(ba *roachpb.BatchRequest) (bool, *roachpb.
 			// has already been finalized, in which case this is a replay.
 			if _, ok := args.(*roachpb.BeginTransactionRequest); ok {
 				key := keys.TransactionKey(header.Key, *ba.GetTxnID())
-				wTS, _, wOK := r.tsCacheMu.cache.GetMaxWrite(key, nil)
+				wTS, _, wOK := r.store.tsCacheMu.cache.GetMaxWrite(key, nil)
 				if wOK {
 					return bumped, roachpb.NewError(roachpb.NewTransactionReplayError())
 				} else if !wTS.Less(ba.Txn.Timestamp) {
@@ -1884,7 +1879,7 @@ func (r *Replica) applyTimestampCache(ba *roachpb.BatchRequest) (bool, *roachpb.
 			}
 
 			// Forward the timestamp if there's been a more recent read (by someone else).
-			rTS, rTxnID, _ := r.tsCacheMu.cache.GetMaxRead(header.Key, header.EndKey)
+			rTS, rTxnID, _ := r.store.tsCacheMu.cache.GetMaxRead(header.Key, header.EndKey)
 			if ba.Txn != nil {
 				if rTxnID == nil || *ba.Txn.ID != *rTxnID {
 					nextTS := rTS.Next()
@@ -1902,7 +1897,7 @@ func (r *Replica) applyTimestampCache(ba *roachpb.BatchRequest) (bool, *roachpb.
 			// write too old boolean for transactions. Note that currently
 			// only EndTransaction and DeleteRange requests update the
 			// write timestamp cache.
-			wTS, wTxnID, _ := r.tsCacheMu.cache.GetMaxWrite(header.Key, header.EndKey)
+			wTS, wTxnID, _ := r.store.tsCacheMu.cache.GetMaxWrite(header.Key, header.EndKey)
 			if ba.Txn != nil {
 				if wTxnID == nil || *ba.Txn.ID != *wTxnID {
 					if !wTS.Less(ba.Txn.Timestamp) {

--- a/pkg/storage/replica_proposal.go
+++ b/pkg/storage/replica_proposal.go
@@ -404,9 +404,16 @@ func (r *Replica) leasePostApply(
 			log.Infof(ctx, "new range lease %s following %s [physicalTime=%s]",
 				newLease, prevLease, r.store.Clock().PhysicalTime())
 		}
-		r.tsCacheMu.Lock()
-		r.tsCacheMu.cache.SetLowWater(newLease.Start)
-		r.tsCacheMu.Unlock()
+		desc := r.Desc()
+		r.store.tsCacheMu.Lock()
+		for _, keyRange := range makeReplicatedKeyRanges(desc) {
+			for _, readOnly := range []bool{true, false} {
+				r.store.tsCacheMu.cache.add(
+					keyRange.start.Key, keyRange.end.Key,
+					newLease.Start, lowWaterTxnIDMarker, readOnly)
+			}
+		}
+		r.store.tsCacheMu.Unlock()
 
 		// Reset the request counts used to make lease placement decisions whenever
 		// starting a new lease.
@@ -422,13 +429,6 @@ func (r *Replica) leasePostApply(
 		}
 	}
 	if leaseChangingHands && !iAmTheLeaseHolder {
-		// We're not the lease holder, reset our timestamp cache, releasing
-		// anything currently cached. The timestamp cache is only used by the
-		// lease holder. Note that we'll call SetLowWater when we next acquire
-		// the lease.
-		r.tsCacheMu.Lock()
-		r.tsCacheMu.cache.Clear(r.store.Clock().Now())
-		r.tsCacheMu.Unlock()
 		// Also clear and disable the push transaction queue. Any waiters
 		// must be redirected to the new lease holder.
 		r.pushTxnQueue.ClearAndDisable()

--- a/pkg/storage/replica_test.go
+++ b/pkg/storage/replica_test.go
@@ -1071,10 +1071,10 @@ func TestReplicaTSCacheLowWaterOnLease(t *testing.T) {
 			t.Fatalf("%d: unexpected error %v", i, err)
 		}
 		// Verify expected low water mark.
-		tc.repl.tsCacheMu.Lock()
-		rTS, _, _ := tc.repl.tsCacheMu.cache.GetMaxRead(roachpb.Key("a"), nil)
-		wTS, _, _ := tc.repl.tsCacheMu.cache.GetMaxWrite(roachpb.Key("a"), nil)
-		tc.repl.tsCacheMu.Unlock()
+		tc.repl.store.tsCacheMu.Lock()
+		rTS, _, _ := tc.repl.store.tsCacheMu.cache.GetMaxRead(roachpb.Key("a"), nil)
+		wTS, _, _ := tc.repl.store.tsCacheMu.cache.GetMaxWrite(roachpb.Key("a"), nil)
+		tc.repl.store.tsCacheMu.Unlock()
 
 		if test.expLowWater == 0 {
 			continue
@@ -1861,28 +1861,28 @@ func TestReplicaUpdateTSCache(t *testing.T) {
 		t.Error(pErr)
 	}
 	// Verify the timestamp cache has rTS=1s and wTS=0s for "a".
-	tc.repl.tsCacheMu.Lock()
-	defer tc.repl.tsCacheMu.Unlock()
-	_, _, rOK := tc.repl.tsCacheMu.cache.GetMaxRead(roachpb.Key("a"), nil)
-	_, _, wOK := tc.repl.tsCacheMu.cache.GetMaxWrite(roachpb.Key("a"), nil)
+	tc.repl.store.tsCacheMu.Lock()
+	defer tc.repl.store.tsCacheMu.Unlock()
+	_, _, rOK := tc.repl.store.tsCacheMu.cache.GetMaxRead(roachpb.Key("a"), nil)
+	_, _, wOK := tc.repl.store.tsCacheMu.cache.GetMaxWrite(roachpb.Key("a"), nil)
 	if rOK || wOK {
 		t.Errorf("expected rOK=false and wOK=false; rOK=%t, wOK=%t", rOK, wOK)
 	}
-	tc.repl.tsCacheMu.cache.ExpandRequests(hlc.Timestamp{})
-	rTS, _, rOK := tc.repl.tsCacheMu.cache.GetMaxRead(roachpb.Key("a"), nil)
-	wTS, _, wOK := tc.repl.tsCacheMu.cache.GetMaxWrite(roachpb.Key("a"), nil)
+	tc.repl.store.tsCacheMu.cache.ExpandRequests(hlc.Timestamp{})
+	rTS, _, rOK := tc.repl.store.tsCacheMu.cache.GetMaxRead(roachpb.Key("a"), nil)
+	wTS, _, wOK := tc.repl.store.tsCacheMu.cache.GetMaxWrite(roachpb.Key("a"), nil)
 	if rTS.WallTime != t0.Nanoseconds() || wTS.WallTime != startNanos || !rOK || wOK {
 		t.Errorf("expected rTS=1s and wTS=0s, but got %s, %s; rOK=%t, wOK=%t", rTS, wTS, rOK, wOK)
 	}
 	// Verify the timestamp cache has rTS=0s and wTS=2s for "b".
-	rTS, _, rOK = tc.repl.tsCacheMu.cache.GetMaxRead(roachpb.Key("b"), nil)
-	wTS, _, wOK = tc.repl.tsCacheMu.cache.GetMaxWrite(roachpb.Key("b"), nil)
+	rTS, _, rOK = tc.repl.store.tsCacheMu.cache.GetMaxRead(roachpb.Key("b"), nil)
+	wTS, _, wOK = tc.repl.store.tsCacheMu.cache.GetMaxWrite(roachpb.Key("b"), nil)
 	if rTS.WallTime != startNanos || wTS.WallTime != t1.Nanoseconds() || rOK || !wOK {
 		t.Errorf("expected rTS=0s and wTS=2s, but got %s, %s; rOK=%t, wOK=%t", rTS, wTS, rOK, wOK)
 	}
 	// Verify another key ("c") has 0sec in timestamp cache.
-	rTS, _, rOK = tc.repl.tsCacheMu.cache.GetMaxRead(roachpb.Key("c"), nil)
-	wTS, _, wOK = tc.repl.tsCacheMu.cache.GetMaxWrite(roachpb.Key("c"), nil)
+	rTS, _, rOK = tc.repl.store.tsCacheMu.cache.GetMaxRead(roachpb.Key("c"), nil)
+	wTS, _, wOK = tc.repl.store.tsCacheMu.cache.GetMaxWrite(roachpb.Key("c"), nil)
 	if rTS.WallTime != startNanos || wTS.WallTime != startNanos || rOK || wOK {
 		t.Errorf("expected rTS=0s and wTS=0s, but got %s %s; rOK=%t, wOK=%t", rTS, wTS, rOK, wOK)
 	}

--- a/pkg/storage/store.go
+++ b/pkg/storage/store.go
@@ -544,6 +544,13 @@ type Store struct {
 		replicaQueues map[roachpb.RangeID]raftRequestQueue
 	}
 
+	tsCacheMu struct {
+		// Protects all fields in the tsCacheMu struct.
+		syncutil.Mutex
+		// Most recent timestamps for keys / key ranges.
+		cache *timestampCache
+	}
+
 	scheduler *raftScheduler
 
 	counts struct {
@@ -918,6 +925,10 @@ func NewStore(cfg StoreConfig, eng engine.Engine, nodeDesc *roachpb.NodeDescript
 	s.mu.replicasByKey = btree.New(64 /* degree */)
 	s.mu.uninitReplicas = map[roachpb.RangeID]*Replica{}
 	s.mu.Unlock()
+
+	s.tsCacheMu.Lock()
+	s.tsCacheMu.cache = newTimestampCache(s.cfg.Clock)
+	s.tsCacheMu.Unlock()
 
 	s.snapshotApplySem = make(chan struct{}, cfg.concurrentSnapshotApplyLimit)
 
@@ -1889,15 +1900,6 @@ func splitPostApply(
 
 	// Finish initialization of the RHS.
 
-	// Copy the timestamp cache from the LHS.
-	// TODO(andrei): We should truncate the entries in both LHS and RHS' timestamp
-	// caches to the respective spans of these new ranges.
-	r.tsCacheMu.Lock()
-	rightRng.tsCacheMu.Lock()
-	r.tsCacheMu.cache.MergeInto(rightRng.tsCacheMu.cache, true /* clear */)
-	rightRng.tsCacheMu.Unlock()
-	r.tsCacheMu.Unlock()
-
 	r.mu.Lock()
 	rightRng.mu.Lock()
 	// Copy the minLeaseProposedTS from the LHS.
@@ -2082,14 +2084,6 @@ func (s *Store) maybeMergeTimestampCaches(
 			"Subsuming lease: %s. Subsumed lease: %s.", subsumingLease, subsumedLease)
 	}
 
-	if subsumingRep.isLeaseValidRLocked(subsumingLease, now) &&
-		subsumingLease.OwnedBy(s.StoreID()) {
-		subsumingRep.tsCacheMu.Lock()
-		subsumedRep.tsCacheMu.Lock()
-		subsumedRep.tsCacheMu.cache.MergeInto(subsumingRep.tsCacheMu.cache, false /* clear */)
-		subsumedRep.tsCacheMu.Unlock()
-		subsumingRep.tsCacheMu.Unlock()
-	}
 	return nil
 }
 

--- a/pkg/storage/timestamp_cache.go
+++ b/pkg/storage/timestamp_cache.go
@@ -20,12 +20,15 @@ import (
 	"fmt"
 	"time"
 
+	"github.com/google/btree"
+	"golang.org/x/net/context"
+
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/util/cache"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/interval"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
-	"github.com/google/btree"
 )
 
 const (
@@ -36,7 +39,13 @@ const (
 	// timestamp.
 	MinTSCacheWindow = 10 * time.Second
 
-	defaultEvictionSizeThreshold = 512
+	// defaultEvictionSizeThreshold is the number of entries to allow in the
+	// per-Store timestamp cache. A cache entry is ~150-200 bytes with the bulk
+	// of that memory coming from the start and end keys.
+	//
+	// TODO(peter): We should accurately track the memory usage of the timestamp
+	// cache and change to using a cache sized by bytes.
+	defaultEvictionSizeThreshold = 1024 * 1024
 
 	// Max entries in each btree node.
 	// TODO(peter): Not yet tuned.
@@ -115,6 +124,21 @@ type timestampCache struct {
 	evictionSizeThreshold int
 }
 
+// lowWaterTxnIDMarker is a special txn ID that identifies a cache entry as a
+// low water mark. It is specified when a lease is acquired to clear the
+// timestamp cache for a range. Also see timestampCache.getMax where this txn
+// ID is checked in order to return whether the max read/write timestamp came
+// from a regular entry or one of these low water mark entries.
+var lowWaterTxnIDMarker = func() *uuid.UUID {
+	// The specific txn ID used here isn't important. We use something that is a)
+	// non-zero and b) obvious.
+	u, err := uuid.FromString("11111111-1111-1111-1111-111111111111")
+	if err != nil {
+		log.Fatal(context.Background(), err)
+	}
+	return &u
+}()
+
 // A cacheValue combines the timestamp with an optional txn ID.
 type cacheValue struct {
 	timestamp hlc.Timestamp
@@ -162,14 +186,6 @@ func (tc *timestampCache) Clear(lowWater hlc.Timestamp) {
 // TimestampCache.
 func (tc *timestampCache) len() int {
 	return tc.rCache.Len() + tc.wCache.Len() + tc.reqSpans
-}
-
-// SetLowWater sets the cache's low water mark, which is the minimum
-// value the cache will return from calls to GetMax().
-func (tc *timestampCache) SetLowWater(lowWater hlc.Timestamp) {
-	if tc.lowWater.Less(lowWater) {
-		tc.lowWater = lowWater
-	}
 }
 
 // add the specified timestamp to the cache as covering the range of
@@ -630,57 +646,10 @@ func (tc *timestampCache) getMax(
 			maxTxnID = nil
 		}
 	}
-	return maxTS, maxTxnID, ok
-}
-
-// MergeInto merges all entries from this timestamp cache into the
-// dest timestamp cache. The clear parameter, if true, copies the
-// values of lowWater and latest and clears the destination cache
-// before merging in the source.
-func (tc *timestampCache) MergeInto(dest *timestampCache, clear bool) {
-	if clear {
-		dest.Clear(tc.lowWater)
-		dest.latest = tc.latest
-		dest.reqIDAlloc = 0
-
-		// Because we just cleared the destination cache, we can directly
-		// insert entries from this cache.
-		hardMerge := func(srcCache, destCache *cache.IntervalCache) {
-			srcCache.Do(func(k, v interface{}) {
-				// Cache entries are mutable (see Add), so we give each cache its own
-				// unique copy.
-				entry := makeCacheEntry(*k.(*cache.IntervalKey), *v.(*cacheValue))
-				destCache.AddEntry(entry)
-			})
-		}
-		hardMerge(tc.rCache, dest.rCache)
-		hardMerge(tc.wCache, dest.wCache)
-	} else {
-		dest.lowWater.Forward(tc.lowWater)
-		dest.latest.Forward(tc.latest)
-
-		// The cache was not cleared before, so we can't just insert entries because
-		// intervals may need to be adjusted or removed to maintain the non-overlapping
-		// guarantee.
-		softMerge := func(srcCache *cache.IntervalCache, readTSCache bool) {
-			srcCache.Do(func(k, v interface{}) {
-				key, val := *k.(*cache.IntervalKey), *v.(*cacheValue)
-				dest.add(roachpb.Key(key.Start), roachpb.Key(key.End), val.timestamp, val.txnID, readTSCache)
-			})
-		}
-		softMerge(tc.rCache, true)
-		softMerge(tc.wCache, false)
+	if maxTxnID == lowWaterTxnIDMarker {
+		ok = false
 	}
-
-	// Copy the requests.
-	tc.requests.Ascend(func(i btree.Item) bool {
-		req := *(i.(*cacheRequest))
-		dest.reqIDAlloc++
-		req.uniqueID = dest.reqIDAlloc
-		dest.requests.ReplaceOrInsert(&req)
-		dest.reqSpans += req.numSpans()
-		return true
-	})
+	return maxTS, maxTxnID, ok
 }
 
 // shouldEvict returns true if the cache entry's timestamp is no

--- a/pkg/storage/timestamp_cache_test.go
+++ b/pkg/storage/timestamp_cache_test.go
@@ -33,7 +33,7 @@ func TestTimestampCache(t *testing.T) {
 	manual := hlc.NewManualClock(baseTS)
 	clock := hlc.NewClock(manual.UnixNano, time.Nanosecond)
 	tc := newTimestampCache(clock)
-	tc.SetLowWater(hlc.Timestamp{WallTime: baseTS})
+	tc.lowWater = hlc.Timestamp{WallTime: baseTS}
 
 	// First simulate a read of just "a" at time 50.
 	tc.add(roachpb.Key("a"), nil, hlc.Timestamp{WallTime: 50}, nil, true)
@@ -99,55 +99,6 @@ func TestTimestampCache(t *testing.T) {
 	}
 }
 
-// TestTimestampCacheSetLowWater verifies that setting the low
-// water mark moves max timestamps forward as appropriate.
-func TestTimestampCacheSetLowWater(t *testing.T) {
-	defer leaktest.AfterTest(t)()
-	manual := hlc.NewManualClock(123)
-	clock := hlc.NewClock(manual.UnixNano, time.Nanosecond)
-	tc := newTimestampCache(clock)
-
-	// Increment time to the low water mark + 10.
-	manual.Increment(10)
-	aTS := clock.Now()
-	tc.add(roachpb.Key("a"), nil, aTS, nil, true)
-
-	// Increment time by 10ns and add another key.
-	manual.Increment(10)
-	bTS := clock.Now()
-	tc.add(roachpb.Key("b"), nil, bTS, nil, true)
-
-	// Increment time by 10ns and add another key.
-	manual.Increment(10)
-	cTS := clock.Now()
-	tc.add(roachpb.Key("c"), nil, cTS, nil, true)
-
-	// Set low water mark.
-	tc.SetLowWater(bTS)
-
-	// Verify looking up key "a" returns the new low water mark ("a"'s timestamp).
-	for i, test := range []struct {
-		key   roachpb.Key
-		expTS hlc.Timestamp
-		expOK bool
-	}{
-		{roachpb.Key("a"), bTS, false},
-		{roachpb.Key("b"), bTS, false},
-		{roachpb.Key("c"), cTS, true},
-		{roachpb.Key("d"), bTS, false},
-	} {
-		if rTS, _, ok := tc.GetMaxRead(test.key, nil); rTS != test.expTS || ok != test.expOK {
-			t.Errorf("%d: expected ts %s, got %s; exp ok=%t; got %t", i, test.expTS, rTS, test.expOK, ok)
-		}
-	}
-
-	// Try setting a lower low water mark than the previous value.
-	tc.SetLowWater(aTS)
-	if rTS, _, ok := tc.GetMaxRead(roachpb.Key("d"), nil); rTS != bTS || ok {
-		t.Errorf("setting lower low water mark should not be allowed; expected %s; got %s; ok=%t", bTS, rTS, ok)
-	}
-}
-
 // TestTimestampCacheEviction verifies the eviction of
 // timestamp cache entries after MinTSCacheWindow interval.
 func TestTimestampCacheEviction(t *testing.T) {
@@ -201,74 +152,6 @@ func TestTimestampCacheNoEviction(t *testing.T) {
 	// Verify that the cache still has 4 entries in it
 	if l, want := tc.len(), 4; l != want {
 		t.Errorf("expected %d entries to remain, got %d", want, l)
-	}
-}
-
-func TestTimestampCacheMergeInto(t *testing.T) {
-	defer leaktest.AfterTest(t)()
-	manual := hlc.NewManualClock(123)
-	clock := hlc.NewClock(manual.UnixNano, time.Nanosecond)
-
-	testCases := []struct {
-		useClear bool
-		expLen   int
-	}{
-		{true, 4},
-		{false, 7},
-	}
-	for i, test := range testCases {
-		tc1 := newTimestampCache(clock)
-		tc2 := newTimestampCache(clock)
-
-		bfTS := clock.Now()
-		tc2.add(roachpb.Key("b"), roachpb.Key("f"), bfTS, nil, true)
-
-		adTS := clock.Now()
-		tc1.add(roachpb.Key("a"), roachpb.Key("d"), adTS, nil, true)
-
-		beTS := clock.Now()
-		tc1.add(roachpb.Key("b"), roachpb.Key("e"), beTS, nil, true)
-
-		aaTS := clock.Now()
-		tc2.add(roachpb.Key("aa"), nil, aaTS, nil, true)
-
-		cTS := clock.Now()
-		tc1.add(roachpb.Key("c"), nil, cTS, nil, true)
-
-		tc1.MergeInto(tc2, test.useClear)
-
-		if tc2.rCache.Len() != test.expLen {
-			t.Errorf("%d: expected merged length of %d; got %d", i, test.expLen, tc2.rCache.Len())
-		}
-		if tc2.latest != tc1.latest {
-			t.Errorf("%d: expected latest to be updated to %s; got %s", i, tc1.latest, tc2.latest)
-		}
-
-		if rTS, _, ok := tc2.GetMaxRead(roachpb.Key("a"), nil); rTS != adTS || !ok {
-			t.Errorf("expected \"a\" to have adTS timestamp; ok=%t", ok)
-		}
-		if rTS, _, ok := tc2.GetMaxRead(roachpb.Key("b"), nil); rTS != beTS || !ok {
-			t.Errorf("expected \"b\" to have beTS timestamp; ok=%t", ok)
-		}
-		if test.useClear {
-			if rTS, _, ok := tc2.GetMaxRead(roachpb.Key("aa"), nil); rTS != adTS || !ok {
-				t.Errorf("expected \"aa\" to have adTS timestamp; ok=%t", ok)
-			}
-		} else {
-			if rTS, _, ok := tc2.GetMaxRead(roachpb.Key("aa"), nil); rTS != aaTS || !ok {
-				t.Errorf("expected \"aa\" to have aaTS timestamp; ok=%t", ok)
-			}
-			if rTS, _, ok := tc2.GetMaxRead(roachpb.Key("a"), roachpb.Key("c")); rTS != aaTS || !ok {
-				t.Errorf("expected \"a\"-\"c\" to have aaTS timestamp; ok=%t", ok)
-			}
-
-			if tc2.latest != cTS {
-				t.Error("expected \"aa\" to have cTS timestamp")
-			}
-			if tc1.latest != cTS {
-				t.Error("expected \"a\"-\"c\" to have cTS timestamp")
-			}
-		}
 	}
 }
 

--- a/pkg/util/cache/cache.go
+++ b/pkg/util/cache/cache.go
@@ -553,11 +553,3 @@ func (ic *IntervalCache) doOverlaps(i interval.Interface) bool {
 	ic.overlaps = append(ic.overlaps, e)
 	return false
 }
-
-// Do invokes f on all of the entries in the cache.
-func (ic *IntervalCache) Do(f func(k, v interface{})) {
-	ic.tree.Do(func(e interval.Interface) (done bool) {
-		f(e.(*Entry).Key, e.(*Entry).Value)
-		return
-	})
-}

